packages/wallets/walletconnect/src/adapter.ts

```diff
@@ -89,74 +105,47 @@ export class WalletConnectWalletAdapter extends BaseSignerWalletAdapter {
-                            resolve(_session);
-                            return;
-                        }
-                    }
-
-                    async function onPairingProposal(proposal: PairingTypes.Proposal) {
-                        const { uri } = proposal.signal.params;
-                        QRCodeModal.open(uri, () => {
-                            cleanup();
-                            reject(new WalletWindowClosedError());
-                        });
-                    }
-
-                    async function onSessionProposal(proposal: SessionTypes.Proposal) {
-                        console.log('onSessionProposal', proposal);
-                    }
-
-                    async function onSessionUpdated(updated: SessionTypes.Settled) {
-                        console.log('onSessionUpdated', updated);
-                        resolve(updated);
-                    }
-
-                    function cleanup() {
-                        client.off(CLIENT_EVENTS.pairing.proposal, onPairingProposal);
-                        client.off(CLIENT_EVENTS.session.proposal, onSessionProposal);
-                        client.off(CLIENT_EVENTS.session.updated, onSessionUpdated);
-                    }
-
-                    try {
-                        client.on(CLIENT_EVENTS.pairing.proposal, onPairingProposal);
-                        client.on(CLIENT_EVENTS.session.proposal, onSessionProposal);
-                        client.on(CLIENT_EVENTS.session.updated, onSessionUpdated);
-
-                        const s = await client.connect(this._params);
-                        resolve(s);
-                    } catch (error: any) {
-                        cleanup();
-                        reject(error);
-                    }
-                });
+                const pairings = client.pairing.getAll({ active: true });
+                // Prototypically, the user should be prompted to either:
+                // - Connect to a previously active pairing
+                // - Choose a new pairing
+                // There doesn't seem to be a WalletConnect-provided UI for this like there exists for the QRCode modal, though,
+                // and pushing this into user-land would be way too much
+                // If we decide to try and pair automatically, the UI will hang waiting for a pairing that might not complete
+                // const lastActivePairing = pairings.length ? pairings[pairings.length - 1].topic : undefined;
+                const lastActivePairing = undefined;
+
+                const { uri, approval } = await client.connect(
+                    getWalletConnectParams(getChainId(this._network), lastActivePairing)
+                );
+
+                if (uri) {
+                    QRCodeModal.open(uri, () => {
+                        throw new WalletWindowClosedError();
+                    });
+                }
+
+                session = await approval();
```

Comment:
This is already much improved! Wondering if we can move more of this adapter's logic into the async loaded library. Check out some other Loadable adapters (e.g. Keystone, Ledger). We moved all complexity out of the adapter itself to keep the footprint and maintenance light.

Reply:
done! check out https://github.com/jnwng/walletconnect-solana/pull/1 (using your fancy-shmancy typescript-monorepo) for all the new logic)